File: Stacks/Models/UserBook.swift
import Foundation

/// UserBook model representing a user's relationship with a book
struct UserBook: Codable, Identifiable {
    let id: Int
    let userID: Int
    let bookID: Int
    var readingStatus: ReadingStatus
    var rating: Int?
    var notes: String?
    var currentPage: Int
    var completionPercentage: Double
    var dateAdded: Date?
    var dateStarted: Date?
    var dateFinished: Date?
    var customMetadata: [String: String]?
    let createdAt: Date
    var updatedAt: Date
    
    // Relationship
    var book: Book?
    
    enum CodingKeys: String, CodingKey {
        case id
        case userID = "user_id"
        case bookID = "book_id"
        case readingStatus = "reading_status"
        case rating
        case notes
        case currentPage = "current_page"
        case completionPercentage = "completion_percentage"
        case dateAdded = "date_added"
        case dateStarted = "date_started"
        case dateFinished = "date_finished"
        case customMetadata = "custom_metadata"
        case createdAt = "created_at"
        case updatedAt = "updated_at"
        case book
    }
}

enum ReadingStatus: String, Codable, CaseIterable {
    case read = "Read"
    case currentlyReading = "Currently Reading"
    case wantToRead = "Want to Read"
    case owned = "Owned"
}

File: backend/src/entities/UserBook.ts
import {
  Entity,
  PrimaryGeneratedColumn,
  Column,
  ManyToOne,
  JoinColumn,
  CreateDateColumn,
} from 'typeorm';
import { User } from './User';
import { Book } from './Book';

export enum BookStatus {
  WANT_TO_READ = 'want_to_read',
  READING = 'reading',
  READ = 'read',
}

@Entity('user_books')
export class UserBook {
  @PrimaryGeneratedColumn()
  id!: number;

  @Column({ name: 'user_id' })
  userId!: number;

  @Column({ name: 'book_id' })
  bookId!: number;

  @Column({
    type: 'enum',
    enum: BookStatus,
    default: BookStatus.WANT_TO_READ,
  })
  status!: BookStatus;

  @CreateDateColumn({ name: 'added_at' })
  addedAt!: Date;

  @ManyToOne(() => User, (user) => user.userBooks)
  @JoinColumn({ name: 'user_id' })
  user!: User;

  @ManyToOne(() => Book, (book) => book.userBooks)
  @JoinColumn({ name: 'book_id' })
  book!: Book;
}

